File: bindings/native/ruby/vibe.c
/*
 * vibe — Ruby C-extension (a TRUE native extension, not fiddle FFI).
 *
 * Built with extconf.rb -> Makefile -> `make`, producing vibe.so which
 * Ruby loads with `require "vibe"`. It #includes the real vibe.h and links
 * libvibe.a, so the Ruby <-> C boundary is compiled native code.
 *
 * Ruby surface:
 *   Vibe.version                     -> String
 *   Vibe.parse(str)                  -> Vibe::Doc      (raises Vibe::Error)
 *   doc.get_string(path)             -> String | nil
 *   doc.get_int(path)                -> Integer
 *   doc.get_float(path)              -> Float
 *   doc.get_bool(path)               -> true/false
 *   doc.array_size(path)             -> Integer
 *   doc.emit                         -> String
 */
#include <ruby.h>
#include <string.h>
#include "vibe.h"

static VALUE mVibe;
static VALUE cDoc;
static VALUE eVibeError;

/* ---- Doc wraps VibeValue* with GC-managed free ------------------------- */
static void doc_free(void *p)
{
    if (p) vibe_value_free((VibeValue *)p);
}

static const rb_data_type_t doc_type = {
    "Vibe::Doc",
    { NULL, doc_free, NULL, },
    NULL, NULL,
    RUBY_TYPED_FREE_IMMEDIATELY,
};

static VibeValue *doc_ptr(VALUE self)
{
    VibeValue *v;
    TypedData_Get_Struct(self, VibeValue, &doc_type, v);
    return v;
}

/* ---- module functions -------------------------------------------------- */
static VALUE vibe_m_version(VALUE self)
{
    return rb_str_new_cstr(vibe_version());
}

static VALUE vibe_m_parse(VALUE self, VALUE rstr)
{
    Check_Type(rstr, T_STRING);
    VibeError err;
    VibeValue *root = vibe_parse(RSTRING_PTR(rstr), (size_t)RSTRING_LEN(rstr), &err);
    if (!root) {
        rb_raise(eVibeError, "%s (line %d, col %d): %s",
                 vibe_error_code_string(err.code), err.line, err.column,
                 err.message ? err.message : "parse error");
    }
    return TypedData_Wrap_Struct(cDoc, &doc_type, root);
}

/* ---- Doc methods ------------------------------------------------------- */
static VALUE doc_get_string(VALUE self, VALUE path)
{
    const char *s = vibe_get_string(doc_ptr(self), StringValueCStr(path));
    return s ? rb_str_new_cstr(s) : Qnil;
}

static VALUE doc_get_int(VALUE self, VALUE path)
{
    return LL2NUM(vibe_get_int(doc_ptr(self), StringValueCStr(path)));
}

static VALUE doc_get_float(VALUE self, VALUE path)
{
    return DBL2NUM(vibe_get_float(doc_ptr(self), StringValueCStr(path)));
}

static VALUE doc_get_bool(VALUE self, VALUE path)
{
    return vibe_get_bool(doc_ptr(self), StringValueCStr(path)) ? Qtrue : Qfalse;
}

static VALUE doc_array_size(VALUE self, VALUE path)
{
    VibeArray *arr = vibe_get_array(doc_ptr(self), StringValueCStr(path));
    return SIZET2NUM(arr ? vibe_array_size(arr) : 0);
}

static VALUE doc_emit(VALUE self)
{
    char *s = vibe_emit(doc_ptr(self));
    if (!s) rb_raise(eVibeError, "vibe_emit failed");
    VALUE out = rb_str_new_cstr(s);
    vibe_free(s);
    return out;
}

void Init_vibe(void)
{
    mVibe = rb_define_module("Vibe");
    eVibeError = rb_define_class_under(mVibe, "Error", rb_eStandardError);

    rb_define_module_function(mVibe, "version", vibe_m_version, 0);
    rb_define_module_function(mVibe, "parse", vibe_m_parse, 1);

    cDoc = rb_define_class_under(mVibe, "Doc", rb_cObject);
    rb_undef_alloc_func(cDoc);
    rb_define_method(cDoc, "get_string", doc_get_string, 1);
    rb_define_method(cDoc, "get_int", doc_get_int, 1);
    rb_define_method(cDoc, "get_float", doc_get_float, 1);
    rb_define_method(cDoc, "get_bool", doc_get_bool, 1);
    rb_define_method(cDoc, "array_size", doc_array_size, 1);
    rb_define_method(cDoc, "emit", doc_emit, 0);
}

File: bindings/native/zig/vibe.zig
//! vibe — native Zig module wrapping libvibe.
//!
//! Unlike an FFI binding, this is a real Zig module you `@import`: it links
//! libvibe.a at build time (see build.zig) and exposes a small typed Zig API
//! over the C ABI (imported via @cImport of the real vibe.h).
//!
//! Build + test:  zig build test   (or: zig build run)

const std = @import("std");

const c = @cImport({
    @cInclude("vibe.h");
});

pub const Error = error{ ParseFailed, EmitFailed };

/// A parsed VIBE document. Call `deinit` to free.
pub const Doc = struct {
    root: *c.VibeValue,

    pub fn parse(data: []const u8) Error!Doc {
        var err: c.VibeError = undefined;
        const root = c.vibe_parse(data.ptr, data.len, &err);
        if (root == null) return Error.ParseFailed;
        return Doc{ .root = root.? };
    }

    pub fn deinit(self: Doc) void {
        c.vibe_value_free(self.root);
    }

    pub fn getString(self: Doc, path: [:0]const u8) ?[]const u8 {
        const s = c.vibe_get_string(self.root, path.ptr);
        if (s == null) return null;
        return std.mem.span(s);
    }

    pub fn getInt(self: Doc, path: [:0]const u8) i64 {
        return c.vibe_get_int(self.root, path.ptr);
    }

    pub fn getFloat(self: Doc, path: [:0]const u8) f64 {
        return c.vibe_get_float(self.root, path.ptr);
    }

    pub fn getBool(self: Doc, path: [:0]const u8) bool {
        return c.vibe_get_bool(self.root, path.ptr);
    }

    pub fn arraySize(self: Doc, path: [:0]const u8) usize {
        const arr = c.vibe_get_array(self.root, path.ptr);
        if (arr == null) return 0;
        return c.vibe_array_size(arr);
    }

    /// Caller owns the returned slice; free with `freeEmit`.
    pub fn emit(self: Doc) Error![]u8 {
        const s = c.vibe_emit(self.root);
        if (s == null) return Error.EmitFailed;
        return std.mem.span(s);
    }
};

pub fn freeEmit(s: []u8) void {
    c.vibe_free(s.ptr);
}

pub fn version() []const u8 {
    return std.mem.span(c.vibe_version());
}

// ---- self-test -------------------------------------------------------------
test "libvibe native module round-trips sample.vibe" {
    const sample = @embedFile("sample.vibe");

    try std.testing.expectEqualStrings("1.1.0", version());

    const doc = try Doc.parse(sample);
    defer doc.deinit();

    try std.testing.expectEqualStrings("libvibe", doc.getString("name").?);
    try std.testing.expectEqual(@as(i64, 42), doc.getInt("answer"));
    try std.testing.expect(@abs(doc.getFloat("pi") - 3.14159) < 1e-9);
    try std.testing.expect(doc.getBool("enabled"));
    try std.testing.expectEqualStrings("localhost", doc.getString("server.host").?);
    try std.testing.expectEqual(@as(i64, 8080), doc.getInt("server.port"));
    try std.testing.expectEqual(@as(usize, 3), doc.arraySize("ports"));

    const out = try doc.emit();
    defer freeEmit(out);
    try std.testing.expect(std.mem.indexOf(u8, out, "libvibe") != null);

    try std.testing.expectError(Error.ParseFailed, Doc.parse("name {"));
}

// A runnable entry point for `zig build run`.
pub fn main() !void {
    const sample = @embedFile("sample.vibe");
    const doc = try Doc.parse(sample);
    defer doc.deinit();
    if (!std.mem.eql(u8, doc.getString("name").?, "libvibe")) return error.Mismatch;
    if (doc.arraySize("ports") != 3) return error.Mismatch;
    const msg = "ALL OK (zig native / module)\n";
    _ = std.c.write(1, msg.ptr, msg.len);
}
